File: ball-bouncing/core/src/com/matthias/Ball.java
package com.matthias;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.GlyphLayout;
import com.badlogic.gdx.math.Circle;
import com.badlogic.gdx.math.Vector2;
import lombok.Getter;

@Getter
public class Ball {

    private static final float BOUNCE_Y_ACCELERATION = 600f;
    private static final float BOUNCE_X_MAX_ACCELERATION = 600f;

    private static final float FALL_ACCELERATION = -350;
    private static final float FALL_VEL_CAP = -500f;

    private final Circle circle;
    private final Color color;
    private final GlyphLayout ballText;

    private float xVel;
    private float yVel;
    private int bounces = 0;

    public Ball(Vector2 pos, float radius, Color color, GlyphLayout ballText) {
        this.circle = new Circle(pos, radius);
        this.color = color;
        this.ballText = ballText;
    }

    public void update(float leftWall, float rightWall, float delta) {
        accelerateFall(delta);
        circle.x += xVel * delta;
        circle.y += yVel * delta;

        bounceLeftWall(leftWall);
        bounceRightWall(rightWall);
    }

    public boolean isInside(Vector2 point) {
        return circle.contains(point);
    }

    public void bounceUp(Vector2 point) {
        if (!isInside(point)) {
            return;
        }

        yVel += BOUNCE_Y_ACCELERATION;
        xVel = BOUNCE_X_MAX_ACCELERATION * (circle.x - point.x) / circle.radius;
        bounces++;
    }

    public boolean isBelowCamera(int cameraFloor) {
        return circle.y + circle.radius < cameraFloor;
    }

    public float getX() {
        return circle.x;
    }

    public float getY() {
        return circle.y;
    }

    public float getRadius() {
        return circle.radius;
    }

    private void accelerateFall(float delta) {
        if (yVel <= FALL_VEL_CAP) {
            yVel = FALL_VEL_CAP;
        } else {
            yVel += FALL_ACCELERATION * delta;
        }
    }

    private void bounceLeftWall(float leftWall) {
        if (circle.x - circle.radius >= leftWall) {
            return;
        }

        xVel *= -1;
        circle.x = leftWall + circle.radius;
        bounces++;

    }

    private void bounceRightWall(float rightWall) {
        if (circle.x + circle.radius <= rightWall) {
            return;
        }

        xVel *= -1;
        circle.x = rightWall - circle.radius;
        bounces++;
    }
}

File: core/src/com/matthias/Ball.kt
package com.matthias

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.GlyphLayout
import com.badlogic.gdx.math.Circle
import com.badlogic.gdx.math.Vector2

class Ball(pos: Vector2?, radius: Float, color: Color, ballText: GlyphLayout) {

    companion object {
        private const val BOUNCE_Y_ACCELERATION = 600f
        private const val BOUNCE_X_MAX_ACCELERATION = 600f
        private const val FALL_ACCELERATION = -350f
        private const val FALL_VEL_CAP = -500f
    }

    private val circle: Circle

    val color: Color
    val ballText: GlyphLayout

    var xVel: Float = 0f
    var yVel: Float = 0f
    var bounces: Int = 0

    init {
        this.circle = Circle(pos, radius)
        this.color = color
        this.ballText = ballText
    }

    fun update(leftWall: Float, rightWall: Float, delta: Float) {
        accelerateFall(delta)
        circle.x += xVel * delta
        circle.y += yVel * delta
        bounceLeftWall(leftWall)
        bounceRightWall(rightWall)
    }

    fun bounceUp(point: Vector2) {
        if (!isInside(point)) {
            return
        }
        yVel += BOUNCE_Y_ACCELERATION
        xVel = BOUNCE_X_MAX_ACCELERATION * (circle.x - point.x) / circle.radius
        bounces++
    }

    fun isBelowCamera(cameraFloor: Int): Boolean {
        return circle.y + circle.radius < cameraFloor
    }

    private fun accelerateFall(delta: Float) {
        if (yVel <= FALL_VEL_CAP) {
            yVel = FALL_VEL_CAP
        } else {
            yVel += FALL_ACCELERATION * delta
        }
    }

    private fun isInside(point: Vector2?): Boolean {
        return circle.contains(point)
    }

    private fun bounceLeftWall(leftWall: Float) {
        if (circle.x - circle.radius >= leftWall) {
            return
        }
        xVel *= -1f
        circle.x = leftWall + circle.radius
        bounces++
    }

    private fun bounceRightWall(rightWall: Float) {
        if (circle.x + circle.radius <= rightWall) {
            return
        }
        xVel *= -1f
        circle.x = rightWall - circle.radius
        bounces++
    }

    val x: Float
        get() = circle.x

    val y: Float
        get() = circle.y

    val radius: Float
        get() = circle.radius
}